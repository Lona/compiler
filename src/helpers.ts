import { IFS } from 'buffs'
import { createModule, ModuleContext } from './logic/module'
import { Config, load } from './utils/config'
import { defaultReporter, Reporter } from './utils/reporter'

/**
 * Helpers passed to every plugins. They contain:
 *
 * - methods abstracting the file system
 * - a centralized log reporter
 * - the workspace's configuration, etc.
 */
export type Helpers = {
  fs: IFS
  reporter: Reporter
  config: Config
  module: ModuleContext
}

export default async (
  fs: IFS,
  workspacePath: string,
  options: {
    outputPath?: unknown
    reporter?: Reporter
  } = {}
): Promise<Helpers> => {
  const { reporter = defaultReporter } = options

  const helpers: Helpers = {
    fs,
    reporter,
    config: await load(fs, workspacePath),
    module: createModule(fs, workspacePath),
  }

  return helpers
}
