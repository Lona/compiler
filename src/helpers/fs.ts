import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
