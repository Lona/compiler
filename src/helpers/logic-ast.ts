import { LogicAST } from '@lona/serialization'
import { uuid } from '../utils'

export { LogicAST as AST }

/**
 * Takes an array of programs and returns a program containing
 * the statements of all programs
 */
export function joinPrograms(
  programs: (LogicAST.Program | void)[]
): LogicAST.Program {
  return {
    type: 'program',
    data: {
      id: uuid(),
      block: programs.reduce(
        (prev, x) => (x ? prev.concat(x.data.block) : prev),
        [] as LogicAST.Statement[]
      ),
    },
  }
}

/** tries to make a program out of any Logic node */
export function makeProgram(
  node: LogicAST.SyntaxNode
): LogicAST.Program | undefined {
  if (node.type === 'program') {
    return node
  }
  if (LogicAST.isStatement(node)) {
    return { type: 'program', data: { id: uuid(), block: [node] } }
  }
  if (LogicAST.isDeclaration(node)) {
    return makeProgram({
      type: 'declaration',
      data: { id: uuid(), content: node },
    })
  }
  if (node.type === 'topLevelDeclarations') {
    return {
      type: 'program',
      data: {
        id: uuid(),
        block: node.data.declarations.map(x => ({
          type: 'declaration',
          data: { id: uuid(), content: x },
        })),
      },
    }
  }

  return undefined
}

export function getPattern(node: LogicAST.SyntaxNode): LogicAST.Pattern | void {
  if (
    node.type === 'variable' ||
    node.type === 'enumeration' ||
    node.type === 'namespace' ||
    node.type === 'record' ||
    node.type === 'importDeclaration' ||
    node.type === 'enumerationCase' ||
    node.type === 'function'
  ) {
    return node.data.name
  }

  if (node.type === 'parameter') {
    if ('localName' in node.data) {
      return node.data.localName
    }
    return node.data.name
  }
}

export function getIdentifier(
  node: LogicAST.SyntaxNode
): LogicAST.Identifier | void {
  if (node.type === 'identifierExpression' || node.type === 'typeIdentifier') {
    return node.data.identifier
  }
  if (node.type === 'memberExpression') {
    return node.data.memberName
  }
}

export function subNodes(node: LogicAST.SyntaxNode): LogicAST.SyntaxNode[] {
  if (node.type === 'loop') {
    return ([node.data.expression] as LogicAST.SyntaxNode[]).concat(
      node.data.block
    )
  }
  if (node.type === 'branch') {
    return ([node.data.condition] as LogicAST.SyntaxNode[]).concat(
      node.data.block
    )
  }
  if (node.type === 'declaration') {
    return [node.data.content]
  }
  if (node.type === 'expression') {
    return [node.data.expression]
  }

  if (node.type === 'variable') {
    return ([] as LogicAST.SyntaxNode[])
      .concat(node.data.annotation ? [node.data.annotation] : [])
      .concat(node.data.initializer ? [node.data.initializer] : [])
  }
  if (node.type === 'function') {
    return ([node.data.returnType] as LogicAST.SyntaxNode[])
      .concat(node.data.genericParameters)
      .concat(node.data.parameters)
      .concat(node.data.block)
  }
  if (node.type === 'enumeration') {
    return ([] as LogicAST.SyntaxNode[])
      .concat(node.data.genericParameters)
      .concat(node.data.cases)
  }
  if (node.type === 'namespace') {
    return node.data.declarations
  }
  if (node.type === 'record') {
    return ([] as LogicAST.SyntaxNode[])
      .concat(node.data.declarations)
      .concat(node.data.genericParameters)
  }
  if (node.type === 'binaryExpression') {
    return [node.data.left, node.data.right, node.data.op]
  }
  if (node.type === 'functionCallExpression') {
    return ([node.data.expression] as LogicAST.SyntaxNode[]).concat(
      node.data.arguments
    )
  }
  if (node.type === 'literalExpression') {
    return [node.data.literal]
  }
  if (node.type === 'memberExpression') {
    return [node.data.expression]
  }

  if (node.type === 'program') {
    return node.data.block
  }

  if (node.type === 'parameter') {
    if ('localName' in node.data) {
      return [node.data.annotation, node.data.defaultValue]
    }
  }

  if (node.type === 'value') {
    return [node.data.expression]
  }

  if (node.type === 'typeIdentifier') {
    return node.data.genericArguments
  }
  if (node.type === 'functionType') {
    return ([node.data.returnType] as LogicAST.SyntaxNode[]).concat(
      node.data.argumentTypes
    )
  }

  if (node.type === 'array') {
    return node.data.value
  }

  if (node.type === 'topLevelParameters') {
    return node.data.parameters
  }

  if (node.type === 'enumerationCase') {
    return node.data.associatedValueTypes
  }

  if (node.type === 'topLevelDeclarations') {
    return node.data.declarations
  }

  if (node.type === 'argument') {
    return [node.data.expression]
  }

  return []
}

export function flattenedMemberExpression(
  memberExpression: LogicAST.Expression
): LogicAST.Identifier[] | void {
  if (memberExpression.type === 'identifierExpression') {
    return [memberExpression.data.identifier]
  }
  if (memberExpression.type !== 'memberExpression') {
    return undefined
  }
  if (memberExpression.data.expression.type === 'identifierExpression') {
    return [
      memberExpression.data.expression.data.identifier,
      memberExpression.data.memberName,
    ]
  }
  const flattenedChildren = flattenedMemberExpression(
    memberExpression.data.expression
  )
  if (!flattenedChildren) {
    return undefined
  }
  return flattenedChildren.concat(memberExpression.data.memberName)
}

export function getNode(
  rootNode: LogicAST.SyntaxNode,
  id: string
): LogicAST.SyntaxNode | undefined {
  if (rootNode.data.id === id) {
    return rootNode
  }

  if ('name' in rootNode.data && rootNode.data.name.id === id) {
    return rootNode
  }

  const children = subNodes(rootNode)

  for (let child of children) {
    const node = getNode(child, id)
    if (node) {
      return node
    }
  }

  return undefined
}

function pathTo(
  rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
  id: string
):
  | (LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier)[]
  | undefined {
  if (id === ('id' in rootNode ? rootNode.id : rootNode.data.id)) {
    return [rootNode]
  }
  if (!('type' in rootNode)) {
    return undefined
  }

  const pattern = getPattern(rootNode)
  if (pattern && pattern.id === id) {
    return [pattern]
  }

  const identifier = getIdentifier(rootNode)
  if (identifier && identifier.id === id) {
    return [identifier]
  }

  return subNodes(rootNode).reduce<
    (LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier)[] | undefined
  >((prev, item) => {
    if (prev) {
      return prev
    }
    const subPath = pathTo(item, id)
    if (subPath) {
      return [rootNode, ...subPath]
    }
    return undefined
  }, undefined)
}

export function findNode(
  rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
  id: string
) {
  const path = pathTo(rootNode, id)
  if (!path) {
    return undefined
  }

  return path[path.length - 1]
}

export function findParentNode(
  node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
  id: string
) {
  const path = pathTo(node, id)
  if (!path || path.length <= 1) {
    return undefined
  }

  const parent = path[path.length - 2]

  if (!('type' in parent)) {
    return undefined
  }

  return parent
}

export function declarationPathTo(
  node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
  id: string
): string[] {
  const path = pathTo(node, id)
  if (!path) {
    return []
  }
  return path.filter(LogicAST.isDeclaration).map(x => {
    switch (x.type) {
      case 'variable':
      case 'function':
      case 'enumeration':
      case 'namespace':
      case 'record':
      case 'importDeclaration': {
        return x.data.name.name
      }
      case 'placeholder': {
        return ''
      }
    }
  })
}
