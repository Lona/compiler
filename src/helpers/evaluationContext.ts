import * as fs from 'fs'
import * as path from 'path'
import { Config } from '../utils/config'
import * as LogicAST from './logicAst'
import * as LogicScope from './logicScope'
import * as LogicUnify from './logicUnify'
import * as LogicEvaluate from './logicEvaluate'
import { Reporter } from '../utils/reporter'
import { nonNullable } from '../utils'

export const STANDARD_LIBRARY = 'standard library'

export const generate = (config: Config, reporter: Reporter) => {
  const standardLibsPath = path.join(__dirname, '../../static/logic')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as LogicAST.AST.Program
  )

  const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

  const logicPrograms = Object.keys(config.logicFiles)
    .map(k => {
      const node = LogicAST.makeProgram(config.logicFiles[k])
      if (!node) {
        return undefined
      }
      return {
        in: k,
        node,
      }
    })
    .filter(nonNullable)

  const scopeContext = LogicScope.build(
    [{ node: standardLibsProgram, in: STANDARD_LIBRARY }].concat(logicPrograms),
    reporter
  )

  const programNode = LogicAST.joinPrograms([
    standardLibsProgram,
    ...logicPrograms.map(x => x.node),
  ])

  const unificationContext = LogicUnify.makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
