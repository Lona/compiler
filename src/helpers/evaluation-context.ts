import * as fs from 'fs'
import * as path from 'path'
import { Config } from '../utils/config'
import * as LogicAST from './logic-ast'
import * as LogicScope from './logic-scope'
import * as LogicUnify from './logic-unify'
import * as LogicEvaluate from './logic-evaluate'
import { Reporter } from './reporter'
import { uuid, nonNullable } from '../utils'

/**
 * Iterates through the import declaration of the program
 * and "resolve them", eg. returns a program that contains both
 * the content of imported files and the current program
 */
function resolveImports(
  program: LogicAST.AST.Program,
  reporter: Reporter,
  existingImports: string[] = []
): LogicAST.AST.Program {
  return {
    type: 'program',
    data: {
      id: uuid(),
      block: program.data.block
        .map(x => {
          if (x.type !== 'declaration') {
            return [x]
          }
          if (x.data.content.type !== 'importDeclaration') {
            return [x]
          }

          const libraryName = x.data.content.data.name.name

          if (existingImports.indexOf(libraryName) !== -1) {
            return [x]
          }

          const libraryPath = path.join(
            __dirname,
            '../../static/logic',
            `${libraryName}.logic`
          )
          const libraryExists = fs.existsSync(libraryPath)

          if (!libraryExists) {
            reporter.error(
              `Failed to find library ${libraryName} at path ${libraryPath}`
            )
            return [x]
          }

          const library = LogicAST.makeProgram(
            JSON.parse(fs.readFileSync(libraryPath, 'utf8'))
          )

          if (!library) {
            reporter.error(`Failed to import library ${libraryName}`)
            return [x]
          }

          const resolvedLibrary = resolveImports(
            library,
            reporter,
            existingImports.concat(libraryName)
          )

          return [x, ...resolvedLibrary.data.block]
        })
        .reduce((prev, x) => prev.concat(x), []),
    },
  }
}

export const generate = (config: Config, reporter: Reporter) => {
  const standardLibsPath = path.join(__dirname, '../../static/logic')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as LogicAST.AST.Program
  )

  const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

  const logicPrograms = Object.keys(config.logicFiles)
    .map(k => {
      const node = LogicAST.makeProgram(config.logicFiles[k])
      if (!node) {
        return undefined
      }
      return {
        in: k,
        node,
      }
    })
    .filter(nonNullable)

  const scopeContext = LogicScope.build(
    [{ node: standardLibsProgram, in: 'standard library' }].concat(
      logicPrograms
    ),
    reporter
  )

  const programNode = LogicAST.joinPrograms([
    standardLibsProgram,
    ...logicPrograms.map(x => x.node),
  ])

  const unificationContext = LogicUnify.makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
