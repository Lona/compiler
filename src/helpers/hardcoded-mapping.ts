// ⚠️ THIS FILE IS AUTO GENERATED. DO NOT MODIFY IT.

import { AST as LogicAST, flattenedMemberExpression } from './logic-ast'
import { EvaluationContext } from './logic-evaluate'

export type HardcodedMap<T, U extends any[]> = {
  functionCallExpression: {
    'Color.setHue': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.setSaturation': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.setLightness': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.fromHSL': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.saturate': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Boolean.or': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Boolean.and': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Number.range': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'String.concat': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Array.at': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Optional.value': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Optional.none': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    Shadow: (node: LogicAST.FunctionCallExpression, ...args: U) => T | undefined
    'FontWeight.ultraLight': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.thin': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.light': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.regular': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.medium': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.semibold': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.bold': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.heavy': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.black': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    TextStyle: (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
  }
  memberExpression: {
    'FontWeight.w100': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w200': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w300': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w400': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w500': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w600': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w700': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w800': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w900': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
  }
}

export const isHardcodedMapCall = {
  functionCallExpression: <T, U extends any[]>(
    k: string,
    map: HardcodedMap<T, U>
  ): k is keyof HardcodedMap<T, U>['functionCallExpression'] => {
    return k in map.functionCallExpression
  },
  memberExpression: <T, U extends any[]>(
    k: string,
    map: HardcodedMap<T, U>
  ): k is keyof HardcodedMap<T, U>['memberExpression'] => {
    return k in map.memberExpression
  },
}

export const createStandardLibraryResolver = <T, U extends any[]>(
  hardcodedMap: HardcodedMap<T, U>
) => (
  node: LogicAST.SyntaxNode,
  evaluationContext: undefined | EvaluationContext,
  ...args: U
): T | undefined => {
  if (!evaluationContext) {
    return undefined
  }

  let matchedHardcodedNode: T | undefined

  if (node.type === 'functionCallExpression') {
    let memberExpression = node.data.expression

    if (!evaluationContext.isFromStandardLibrary(memberExpression.data.id)) {
      return
    }

    const path = (flattenedMemberExpression(memberExpression) || [])
      .map(y => y.string)
      .join('.')

    if (isHardcodedMapCall.functionCallExpression(path, hardcodedMap)) {
      matchedHardcodedNode = hardcodedMap.functionCallExpression[path](
        node,
        ...args
      )
    }
  }

  if (node.type === 'memberExpression') {
    let memberExpression = node

    if (!evaluationContext.isFromStandardLibrary(memberExpression.data.id)) {
      return
    }

    const path = (flattenedMemberExpression(memberExpression) || [])
      .map(y => y.string)
      .join('.')

    if (isHardcodedMapCall.memberExpression(path, hardcodedMap)) {
      matchedHardcodedNode = hardcodedMap.memberExpression[path](node, ...args)
    }
  }

  return matchedHardcodedNode
}
