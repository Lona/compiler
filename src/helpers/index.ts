import { LogicAST } from '@lona/serialization'
import { config as Config } from '../utils'
import { generate as generateEvaluationContext } from './evaluation-context'
import { EvaluationContext } from './logic-evaluate'
import { createStandardLibraryResolver } from './hardcoded-mapping'
import { Reporter, defaultReporter } from './reporter'
import { FSWrapper, createFSWrapper } from './fs'
export { HardcodedMap } from './hardcoded-mapping'
export { EvaluationContext } from './logic-evaluate'
import { TraversalConfig, reduce } from './logic-traversal'
import {
  makeProgram,
  declarationPathTo,
  findParentNode,
  findNode,
} from './logic-ast'

/**
 * Helpers passed to every plugins. They contains some methods abstracting
 * the file system, the evaluation context of the workspace,
 * a reporter (to centralize the logs), the workspace's configuration, etc.
 */
export type Helpers = {
  fs: FSWrapper
  reporter: Reporter
  config: Config.Config
  /**
   * The evaluation context of the Lona Workspace.
   *
   * It contains some mapping between the identifiers and what they reference,
   * eg. `Optional.none` -> the function declaration in Prelude.logic
   * which is used to evaluate `Optional.none` to
   * `{ type: unit, memory: { type: 'unit' } }` for example. It takes care of
   * the scope of the identifier to determine the reference.
   *
   * Additionally, it keeps track of where declaration are from.
   */
  evaluationContext: EvaluationContext | undefined
  /**
   * Takes a hardcoded map `standard library -> node` and creates a function
   * which returns a node if the Logic node passed as argument is from
   * the standard library.
   *
   * Depending on the context, the hardcoded map might use the
   * evaluation context to evaluate the Logic node instead of trying to translate
   * the standard library method into the target format.
   */
  createStandardLibraryResolver: typeof createStandardLibraryResolver
  ast: {
    /** Tries to make a program out of any Logic node */
    makeProgram: (node: LogicAST.SyntaxNode) => LogicAST.Program | undefined
    /** A set of methods to help traverse the AST */
    traversal: {
      declarationPathTo: (
        node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => string[]
      /** Find the node with the matching id */
      findNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) =>
        | LogicAST.SyntaxNode
        | LogicAST.Pattern
        | LogicAST.Identifier
        | undefined
      /** Find the parent of the node with the matching id */
      findParentNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => LogicAST.SyntaxNode | undefined
      /**
       * The `reduce()` method executes a reducer function (that you provide)
       * on each node of the AST, resulting in a single output value.
       *
       * Your reducer function's returned value is assigned to the accumulator,
       * whose value is remembered across each iteration throughout the AST,
       * and ultimately becomes the final, single resulting value.
       *
       * The traversal is depth-first, you can specify whether it should be
       * pre-order or post-order in the config.
       */
      reduce: <T>(
        node: LogicAST.SyntaxNode,
        callbackfn: (
          previousValue: T,
          currentNode: LogicAST.SyntaxNode,
          config: TraversalConfig
        ) => T,
        initialResult: T,
        config?: TraversalConfig
      ) => T
    }
  }
}

export type PreludeFlags = {
  [id: string]:
    | LogicAST.RecordDeclaration
    | LogicAST.EnumerationDeclaration
    | LogicAST.NamespaceDeclaration
}

export default async (
  workspacePath: string,
  outputPath?: unknown,
  _reporter?: Reporter
): Promise<Helpers> => {
  const fsWrapper = createFSWrapper(workspacePath, outputPath)

  const config = await Config.load(workspacePath, {
    forEvaluation: true,
    fs: fsWrapper,
  })

  let cachedEvaluationContext: EvaluationContext | undefined

  const reporter = _reporter || defaultReporter

  return {
    fs: fsWrapper,
    config,
    get evaluationContext() {
      if (cachedEvaluationContext) {
        return cachedEvaluationContext
      }
      cachedEvaluationContext = generateEvaluationContext(config, reporter)
      return cachedEvaluationContext
    },
    createStandardLibraryResolver,
    reporter,
    ast: {
      makeProgram,
      traversal: {
        declarationPathTo,
        findNode,
        findParentNode,
        reduce,
      },
    },
  }
}
