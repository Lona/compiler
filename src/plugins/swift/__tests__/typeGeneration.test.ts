import { createFs } from 'buffs'
import fs from 'fs'
import path from 'path'
import { createHelpers } from '../../../helpers'
import plugin from '../index'

it('converts Swift language', async () => {
  const source = createFs({
    'lona.json': JSON.stringify({}),
    'SwiftLanguage.logic': fs.readFileSync(
      path.join(__dirname, './mocks/SwiftLanguage.logic'),
      'utf8'
    ),
  })

  const helpers = createHelpers(source, '/')

  await plugin.convertWorkspace('/', helpers, { output: '/output' })

  const colors = source.readFileSync('/output/SwiftLanguage.swift', 'utf8')

  expect(colors).toMatchSnapshot()
})
