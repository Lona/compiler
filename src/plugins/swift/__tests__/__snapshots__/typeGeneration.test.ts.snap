// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts Swift language 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public indirect enum AccessLevelModifier {
  case privateModifier
  case fileprivateModifier
  case internalModifier
  case publicModifier
  case openModifier
}
public indirect enum MutationModifier {
  case mutatingModifier
  case nonmutatingModifier
}
public indirect enum OtherModifier {
  case classModifier
  case convenienceModifier
  case dynamicModifier
  case finalModifier
  case infixModifier
  case lazyModifier
  case optionalModifier
  case overrideModifier
  case postfixModifier
  case prefixModifier
  case requiredModifier
  case staticModifier
  case unownedModifier
  case unownedSafeModifier
  case unownedUnsafeModifier
  case weakModifier
}
public indirect enum Literal {
  case nil
  case boolean(Bool)
  case integer(CGFloat)
  case floatingPoint(CGFloat)
  case string(String)
  case color(String)
  case image(String)
  case array(Array<Literal>)
}
"
`;
